### ISSUE TITLE
hs.json.decode bug (introduced in 0.9.85?)

### ISSUE THREAD

**Issue by smackesey** [Hammerspoon/hammerspoon#2778]
I have a bunch of code that relies on `hs.json.decode` that stopped working after the 0.9.85 update (still doesn't work in 0.9.86). I investigated a bit and couldn't figure out what was going on-- `hs.json.decode` seemed to sometimes be returning a string when it should've been returning a table. I was observing this by inserting `type(hs.json.decode(myJsonStr))` (which would return 'string' when it should've returned 'table') within the functions I was using, but I couldn't replicate this in the console.

In any case, I've temporarily fixed my issues by installing https://github.com/neoxic/lua-json and overwriting hs.json.decode in my init.lua:

```lua
local js = require('json')

local function decodeJson(jsonStr)
  local decoded, _ = js.decode(jsonStr)
  decoded['__array'] = nil
  return decoded
end

hs.json.decode = decodeJson
```

**Comment by latenitefilms** [comment]
@cmsj recently modernised `hs.json`, so something must have changed in the process. I can't spot any obvious bugs in the current code, but will take a deeper look tomorrow.

**Comment by cmsj** [comment]
@smackesey do you happen to have some json that triggers the buggy behaviour?

**Comment by smackesey** [comment]
@cmsj Yes, but the code is not standalone. Here's the function where I first discovered the bug.

```lua
function m.getWindows(spaceRep)
  local windows
  if spaceRep then
    local spaceIndex = normalize.space(spaceRep, 'index')
    windows = m.command(string.format('query --windows --space %d', spaceIndex))
  else
    windows = m.command('query --windows')
  end
  return hs.json.decode(windows)
end
```

`m.command` here is executing a `yabai` command via coroutines and `hs.task`, and returns a json string.

**Comment by cmsj** [comment]
@latenitefilms I wonder if this is because we're taking a string from Lua, converting it to an NSString, then converting it to an NSData to feed to NSJSONSerialization - maybe it would be better to do the initial `toNSObjectAtIndex` with the `LS_NSLuaStringAsDataOnly` option and pass that directly to NSJSONSerialization.

**Comment by latenitefilms** [comment]
@smackesey - are you able to add in some debugging code so we can try and work out why/where it's breaking?

```lua
function m.getWindows(spaceRep)
  local windows
  if spaceRep then
    local spaceIndex = normalize.space(spaceRep, 'index')
    windows = m.command(string.format('query --windows --space %d', spaceIndex))
  else
    windows = m.command('query --windows')
  end
  local result = hs.json.decode(windows)
  if type(result) == "string" then
    print(string.format("this shouldn't be a string: %s", result))
    print(string.format("windows: %s", hs.inspect(windows)))
  end 
  return result 
end
```

**Comment by smackesey** [comment]
Hey, I saw that this was announced fixed in the 0.9.87 release, but the problem still persists on my end. I updated to 0.9.87 a few weeks ago but just now removed the patch in the OP, and I'm getting the same error. Apologies @latenitefilms for not responding to your earlier request for debugging assistance. Happy to help any way I can now to get this fixed. Perhaps it's related to the newer issue someone posted about `hs.json` (#2826).
